pkg/webhook/strategy/webhook.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2019. PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package strategy
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// AdmissionWebhook is a admission webhook based on the registered strategies in the given registry
+type AdmissionWebhook struct {
+	registry *StrategyRegistry
+}
+
+func NewAdmissionWebhook(registry *StrategyRegistry) *AdmissionWebhook {
+	return &AdmissionWebhook{registry}
+}
+
+func (w *AdmissionWebhook) Validate(ar *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
```

Comment:
can we have some basic tests on `Validation/Mutate` methods? it helps readers to understand `AdmissionRequest` and `AdmissionResponse `, more cases can be added in later development.

Reply:
sure

Reply:
Addressed in https://github.com/pingcap/tidb-operator/pull/1429/commits/60d3c5bf860ebba5a8877b529d633439a199a28d